### PR TITLE
Add pods stability for health check.

### PR DESCRIPTION
**Pull request by tejal29**
This PR checks if Pods are either running or succeed whichever event is seen  first within 5 minutes.
More context https://github.com/GoogleContainerTools/skaffold/pull/2359

Add pods Stability 

An example output will look like this for a number of pods:
```

tejaldesai:unstable-deployment (add_pods_stability)$ skaffold deploy
Starting deploy...
kubectl client version: 1.11+
kubectl version 1.12.0 or greater is recommended for use with Skaffold
deployment.apps/unstable-deployment unchanged
Deploy complete in 791.09378ms
Waiting for deployments to stabilize
Waiting on 2 of 2 deployments
Waiting on 5 of 5 pods
Waiting on 4 of 5 pods
Waiting on 3 of 5 pods
Waiting on 2 of 5 pods
Waiting on 1 of 5 pods
Waiting on 0 of 5 pods
Waiting on 1 of 2 deployments
Waiting on 0 of 2 deployments
following resources are not stable:
deployment/leeroy-web failed due to Running [kubectl --context gke_tejal-test_us-central1-a_dump rollout status deployment leeroy-web --watch=false]: stdout , stderr: Error from server (NotFound): deployments.extensions "leeroy-web" not found
, err: exit status 1: exit status 1
deployment/unstable-deployment failed due to Running [kubectl --context gke_tejal-test_us-central1-a_dump rollout status deployment unstable-deployment --watch=false]: stdout , stderr: error: deployment "unstable-deployment" exceeded its progress deadline
, err: exit status 1: exit status 1
FATA[0001] following resources are not stable:
deployment/leeroy-web failed due to Running [kubectl --context gke_tejal-test_us-central1-a_dump rollout status deployment leeroy-web --watch=false]: stdout , stderr: Error from server (NotFound): deployments.extensions "leeroy-web" not found
, err: exit status 1: exit status 1
deployment/unstable-deployment failed due to Running [kubectl --context gke_tejal-test_us-central1-a_dump rollout status deployment unstable-deployment --watch=false]: stdout , stderr: error: deployment "unstable-deployment" exceeded its progress deadline
, err: exit status 1: exit status 1 
tejaldesai:unstable-deployment (add_pods_stability)$ 
```

### CODE REVIEW

**Comment by priyawadhwa** [comment]
Looks like a failure in the deploy unit test

```
--- FAIL: TestDeploy (0.01s)
    --- FAIL: TestDeploy/deploy_shd_perform_status_check (0.00s)
        util.go:68: unexpected error: getting client config for kubernetes client: error creating kubeConfig: invalid configuration: no configuration has been provided
```

File: pkg/skaffold/deploy/status_check.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+func podErrors(m *sync.Map) error {
+	errorStrings := []string{}
+	m.Range(func(k, v interface{}) bool {
+		if _, ok := v.(error); ok {
+			errorStrings = append(errorStrings, fmt.Sprintf("pod %s is not stable", k))
+		}
+		return true
+	})
+
+	if len(errorStrings) == 0 {
```
Comment (priyawadhwa) [inline pkg/skaffold/deploy/status_check.go]: i think this check should happen first?

File: pkg/skaffold/runner/runner.go [hunk 2]
```diff
@@ -218,14 +222,18 @@ func (r *SkaffoldRunner) Deploy(ctx context.Context, out io.Writer, artifacts []
...
 
-func (r *SkaffoldRunner) performStatusCheck(out io.Writer) error {
+func (r *SkaffoldRunner) performStatusCheck(ctx context.Context, out io.Writer) error {
 	// Check if we need to perform deploy status
 	if r.runCtx.Opts.StatusCheck {
 		fmt.Fprintln(out, "Waiting for deployments to stabilize")
-		// TODO : Actually perform status check
+		err := statusCheck(ctx, r.defaultLabeller, r.runCtx)
+		if err != nil {
+			fmt.Fprintln(out, err.Error())
```
Comment (priyawadhwa) [inline pkg/skaffold/runner/runner.go]: will this error be printed twice if we `return err` again next?

**Comment by tejal29** [comment]
> Looks like a failure in the deploy unit test
> 
> ```
> --- FAIL: TestDeploy (0.01s)
>     --- FAIL: TestDeploy/deploy_shd_perform_status_check (0.00s)
>         util.go:68: unexpected error: getting client config for kubernetes client: error creating kubeConfig: invalid configuration: no configuration has been provided
> ```

Looks like this is a flake. it passes locally on my branch. I will try running it several times.

File: pkg/skaffold/deploy/status_check.go (hunk 1 above)
Comment (tejal29) [reply pkg/skaffold/deploy/status_check.go]: i traverse through the sync map and collect errors in `errorStrings` in the lines above.  If the `return true` on L90 is what is throwing you off, its returning from the map traverse function.

File: pkg/skaffold/runner/runner.go (hunk 2 above)
Comment (tejal29) [reply pkg/skaffold/runner/runner.go]: no, it does not. 
https://github.com/GoogleContainerTools/skaffold/blob/master/cmd/skaffold/app/cmd/deploy.go#L55

```
func doDeploy(ctx context.Context, out io.Writer) error {
	return withRunner(ctx, func(r runner.Runner, _ *latest.SkaffoldConfig) error {
		// If the BuildArtifacts contains an image in the preBuilt list,
		// use image from BuildArtifacts instead
		deployArtifacts := build.MergeWithPreviousBuilds(buildOutputFile.BuildArtifacts(), preBuiltImages.Artifacts())

		return r.DeployAndLog(ctx, out, deployArtifacts)
	})
}
```
And DeployAndLog, does print it out either
```
func (r *SkaffoldRunner) DeployAndLog(ctx context.Context, out io.Writer, artifacts []build.Artifact) error {
	if !r.runCtx.Opts.Tail {
		return r.Deploy(ctx, out, artifacts)
	}
	....
	if err := r.Deploy(ctx, out, artifacts); err != nil {
		return err
	}
```

I thought about printing the error in cmd/skaffold/app/cmd/deploy.go but its called a lot of places and did not want to repeat it everywhere.

Comment (priyawadhwa) [reply pkg/skaffold/runner/runner.go]: makes sense! i think this answers a comment I made on your other PR as well :sweat_smile:

File: pkg/skaffold/deploy/status_check.go [hunk 3]
```diff
@@ -40,45 +44,76 @@ var (
...
+	defer wg.Wait()
 	kubeCtl := &kubectl.CLI{
 		Namespace:   runCtx.Opts.Namespace,
 		KubeContext: runCtx.KubeContext,
 	}
+	numDeps := int32(len(dMap))
+	var ops int32
+	atomic.StoreInt32(&ops, numDeps)
 
+	fmt.Fprintln(out, fmt.Sprintf("Waiting on %d of %d deployments", atomic.LoadInt32(&ops), numDeps))
```
Comment (balopat) [inline pkg/skaffold/deploy/status_check.go]: use `color.Default.Frpintf` instead

File: pkg/skaffold/deploy/status_check.go [hunk 4]
```diff
@@ -40,45 +44,76 @@ var (
...
 	for dName, deadline := range dMap {
 		deadlineDuration := time.Duration(deadline) * time.Second
 		wg.Add(1)
 		go func(dName string, deadlineDuration time.Duration) {
-			defer wg.Done()
+			defer func() {
+				atomic.AddInt32(&ops, -1)
+				fmt.Fprintln(out, fmt.Sprintf("Waiting on %d of %d deployments", atomic.LoadInt32(&ops), numDeps))
```
Comment (balopat) [inline pkg/skaffold/deploy/status_check.go]: use `color.Default.Frpintf` instead

File: pkg/skaffold/deploy/status_check.go [hunk 5]
```diff
@@ -111,39 +146,80 @@ func pollDeploymentRolloutStatus(ctx context.Context, k *kubectl.CLI, dName stri
...
+
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+	fmt.Fprintln(out, fmt.Sprintf("Waiting on %d of %d pods", atomic.LoadInt32(&ops), numPods))
+	for _, po := range pods {
+		wg.Add(1)
+		go func(po *v1.Pod) {
+			defer func() {
+				atomic.AddInt32(&ops, -1)
+				fmt.Fprintln(out, fmt.Sprintf("Waiting on %d of %d pods", atomic.LoadInt32(&ops), numPods))
```
Comment (balopat) [inline pkg/skaffold/deploy/status_check.go]: use `color.Default.Frpintf` instead

File: integration/testdata/unstable-deployment/incorrect-deployment.yaml [hunk 6]
```diff
@@ -15,9 +15,9 @@ spec:
     spec:
       containers:
         - name: incorrect-example
-          image: gcr.io/k8s-skaffold/incorrect-example
+          image: gcr.io/k8s-skaffold/skaffold-example
```
Comment (balopat) [inline integration/testdata/unstable-deployment/incorrect-deployment.yaml]: why?

**Comment by balopat** [comment]
4. It's good to have the number of pods that we are waiting still, but when we _know_ that a pod/deployment is errored/succeeded print that out as well. `pod foobar is Ready (4/13 pods still waiting)` or  `pod foobar Failed due to 'PodUnschedulable blablabla' (4/13 pods still waiting)`

**Comment by tejal29** [comment]
@balopat Please take a look at the new changes.
I tested this on microservices/demo here with few changes. 
https://github.com/tejal29/microservices-demo/tree/healthcheck_demo

**Comment by tejal29** [comment]
@balopat please take a look at the final output
https://gist.github.com/tejal29/02bab3f291d7fdc80066a0d825b4b1a2

**Comment by tejal29** [comment]
Will wait for #2591

**Comment by nkubala** [comment]
@tejal29 since #2591 is now in can you get this one rebased?

**Comment by tejal29** [comment]
Discard in favour of #2750

**Pull request closed without merging**